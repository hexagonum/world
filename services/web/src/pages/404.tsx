import ErrorTemplate from '@world/templates/ErrorTemplate';
import { NextPage } from 'next';

export const NotFoundPage: NextPage = () => (
  <ErrorTemplate code={404} message="Page Not Found" />
);

export default NotFoundPage;
