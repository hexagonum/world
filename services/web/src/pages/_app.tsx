import { ApolloProvider } from '@apollo/client';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import { ChakraProvider } from '@chakra-ui/react';
import { APP_NAME } from '@world/common/constants';
import { apolloClient } from '@world/graphql';
import '@world/styles/globals.scss';
import type { AppProps } from 'next/app';
import Head from 'next/head';

const App: React.FC<AppProps> = ({ Component, pageProps }: AppProps) => {
  return (
    <>
      <Head>
        <title>{APP_NAME}</title>
      </Head>
      <UserProvider>
        <ApolloProvider client={apolloClient}>
          <ChakraProvider>
            <Component {...pageProps} />
          </ChakraProvider>
        </ApolloProvider>
      </UserProvider>
    </>
  );
};

export default App;
