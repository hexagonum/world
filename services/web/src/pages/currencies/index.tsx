import { Table, TableCaption, TableContainer, Tbody, Td, Th, Thead, Tr } from '@chakra-ui/react';
import Container from '@world/components/Container';
import currencies from '@world/data/currencies/list.json';
import useFetch from '@world/hooks/use-fetch';
import Layout from '@world/layout';
import { currencyFormatter } from '@world/utils/currency-formatter';
import { NextPage } from 'next';
import Link from 'next/link';

type CurrenciesListProps = {
  loading?: boolean;
  error?: Error;
  data?: RatesResponse;
};

const CurrenciesList: React.FC<CurrenciesListProps> = ({ loading, error, data }) => {
  if (loading) {
    return <div className="flex justify-center items-center py-8">Loading</div>;
  }

  if (error) {
    return <div className="flex justify-center items-center py-8">{error.message}</div>;
  }

  if (!data) {
    return <div className="flex justify-center items-center py-8">No Data</div>;
  }

  const { rates = {}, base = '' } = data;

  return (
    <TableContainer className="border rounded shadow">
      <Table>
        <Thead>
          <Tr>
            <Th>Currencies ({currencies.length})</Th>
            <Th isNumeric>Countries</Th>
            <Th isNumeric>Rate</Th>
          </Tr>
        </Thead>
        <Tbody>
          {currencies.map(({ code = '', name = '', total = 0 }) => {
            let rate: string = 'N/A';
            if (base === code) {
              rate = '1';
            } else if (rates[code]) {
              rate = currencyFormatter(rates[code], code);
            }
            return (
              <Tr key={code}>
                <Td>
                  <Link href={`/currencies/${code}`}>{name}</Link>
                </Td>
                <Td isNumeric>{total}</Td>
                <Td isNumeric>{rate}</Td>
              </Tr>
            );
          })}
        </Tbody>
        <TableCaption>
          <p className="pb-4">Currencies ({currencies.length})</p>
        </TableCaption>
      </Table>
    </TableContainer>
  );
};

type RatesResponse = {
  amount: number;
  base: string;
  date: string;
  rates: Record<string, number>;
};

const CurrenciesPage: NextPage = () => {
  const base: string = 'https://api.frankfurter.app';
  const url: string = `${base}/latest?base=EUR&amount=${1}`;
  const { data, loading, error } = useFetch<RatesResponse>(url);

  return (
    <Layout>
      <Container>
        <div className="p-8">
          <CurrenciesList loading={loading} error={error} data={data} />
        </div>
      </Container>
    </Layout>
  );
};

export default CurrenciesPage;
