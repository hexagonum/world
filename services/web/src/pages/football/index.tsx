import { Input, Table, TableCaption, TableContainer, Tbody, Td, Th, Thead, Tr } from '@chakra-ui/react';
import Container from '@world/components/Container';
import { apolloClient } from '@world/graphql';
import { FOOTBALL_AREAS_QUERY } from '@world/graphql/queries/football';
import Layout from '@world/layout';
import { FootballArea } from '@world/types/football';
import { NextPage } from 'next';
import Link from 'next/link';
import { ChangeEvent, useState } from 'react';

export const FootballPage: NextPage<{ areas: FootballArea[] }> = ({ areas = [] }) => {
  const [query, setQuery] = useState<string>('');

  const filteredAreas = areas.filter(({ name }) => {
    const nameFlag = query !== '' ? name.toLowerCase().includes(query.toLowerCase()) : true;
    return nameFlag;
  });

  return (
    <Layout
      searchSection={
        <>
          <Input
            id="query"
            name="query"
            placeholder="Query"
            value={query}
            onChange={(event: ChangeEvent<HTMLInputElement>) => setQuery(event.target.value)}
            className="shadow"
          />
        </>
      }
    >
      <Container>
        <div className="p-8">
          <div className="flex flex-col gap-4 md:gap-8">
            <TableContainer className="border rounded shadow">
              <Table>
                <Thead>
                  <Tr>
                    <Th>Name</Th>
                  </Tr>
                </Thead>
                <Tbody>
                  {filteredAreas.map(({ id, name }) => {
                    return (
                      <Tr key={id}>
                        <Td>
                          <Link href={`/football/${id}`}>{name}</Link>
                        </Td>
                      </Tr>
                    );
                  })}
                </Tbody>
                <TableCaption>
                  <p className="pb-4">Areas ({areas.length})</p>
                </TableCaption>
              </Table>
            </TableContainer>
          </div>
        </div>
      </Container>
    </Layout>
  );
};

export const getStaticProps = async (): Promise<{ props: { areas: FootballArea[] } }> => {
  try {
    const data = await apolloClient.query<{ football: { areas: FootballArea[] } }>({ query: FOOTBALL_AREAS_QUERY });
    const areas = [...data.data.football.areas];
    return { props: { areas } };
  } catch (error) {
    console.error(error);
    return { props: { areas: [] } };
  }
};

export default FootballPage;
