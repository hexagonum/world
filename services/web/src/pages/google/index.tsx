import Container from '@world/components/Container';
import Layout from '@world/layout';
import { NextPage } from 'next';

const GooglePage: NextPage = () => {
  return (
    <Layout>
      <Container>
        <div className="p-8"></div>
      </Container>
    </Layout>
  );
};

export default GooglePage;
