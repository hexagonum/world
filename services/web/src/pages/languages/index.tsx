import { Table, TableContainer, Tbody, Td, Th, Thead, Tr } from '@chakra-ui/react';
import Container from '@world/components/Container';
import { NEXT_PUBLIC_BASE_API } from '@world/configs';
import Layout from '@world/layout';
import { NextPage } from 'next';
import Link from 'next/link';

type Language = {
  code: string;
  name: string;
  countries: { commonName: string; region: string; subregion: string; population: number }[];
};

type LanguagesPageProps = {
  languages: Language[];
};

const LanguagesPage: NextPage<LanguagesPageProps> = ({ languages = [] }) => {
  return (
    <Layout>
      <Container>
        <div className="p-8">
          <TableContainer className="border rounded shadow">
            <Table>
              <Thead>
                <Tr>
                  <Th>Languages ({languages.length})</Th>
                  <Th isNumeric>Countries</Th>
                  <Th isNumeric>Population</Th>
                </Tr>
              </Thead>
              <Tbody>
                {languages.map(({ code = '', name = '', countries = [] }) => {
                  const population: number = countries
                    .map(({ population = 0 }) => population)
                    .reduce((previousValue: number, currentValue: number) => previousValue + currentValue, 0);

                  return (
                    <Tr key={code}>
                      <Td>
                        <Link href={`/languages/${code}`}>{name}</Link>
                      </Td>
                      <Td isNumeric>{countries.length}</Td>
                      <Td isNumeric>{population.toLocaleString()}</Td>
                    </Tr>
                  );
                })}
              </Tbody>
            </Table>
          </TableContainer>
        </div>
      </Container>
    </Layout>
  );
};

export const getStaticProps = async (): Promise<{ props: { languages: Language[] } }> => {
  try {
    const response = await fetch(`${NEXT_PUBLIC_BASE_API}/languages`);
    const languages: Language[] = await response.json();
    return { props: { languages } };
  } catch (error) {
    console.error(error);
    return { props: { languages: [] } };
  }
};

export default LanguagesPage;
