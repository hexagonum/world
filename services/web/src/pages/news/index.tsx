import { Button } from '@chakra-ui/react';
import Container from '@world/components/Container';
import { NEXT_PUBLIC_BASE_API } from '@world/configs';
import { Articles, Article } from '@world/components/Articles';
import { Layout } from '@world/layout';
import { NextPage } from 'next';
import { useState } from 'react';

type NewsPageProps = {
  articles: Article[];
};

const categories = ['general', 'business', 'entertainment', 'health', 'science', 'sports', 'technology'];

const NewsPage: NextPage<NewsPageProps> = ({ articles = [] }) => {
  const [category, setCategory] = useState('general');

  return (
    <Layout>
      <Container>
        <div className="p-8">
          <div className="flex flex-col gap-4 md:gap-8">
            <div className="w-full overflow-auto">
              <div className="flex gap-4">
                {categories.map((c) => (
                  <div key={c}>
                    <Button
                      colorScheme="teal"
                      className="capitalize"
                      variant={c === category ? 'solid' : 'outline'}
                      onClick={() => setCategory(c)}
                    >
                      {c}
                    </Button>
                  </div>
                ))}
              </div>
            </div>
            <Articles category={category} country="us" articles={articles} />
          </div>
        </div>
      </Container>
    </Layout>
  );
};

export const getStaticProps = async (): Promise<{ props: { articles: Article[] } }> => {
  try {
    const response = await fetch(`${NEXT_PUBLIC_BASE_API}/news/headlines?country=us`);
    const articles: Article[] = await response.json();
    return { props: { articles } };
  } catch (error) {
    console.error(error);
    return { props: { articles: [] } };
  }
};

export default NewsPage;
