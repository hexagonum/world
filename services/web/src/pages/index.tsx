import {
  Badge,
  Button,
  Card,
  CardBody,
  CardHeader,
  Divider,
  Table,
  TableCaption,
  TableContainer,
  Tbody,
  Td,
  Text,
  Th,
  Thead,
  Tr,
} from '@chakra-ui/react';
import Container from '@world/components/Container';
import { Weather } from '@world/components/Weather';
import { apolloClient } from '@world/graphql';
import { HOME_QUERY } from '@world/graphql/queries/home';
import { Layout } from '@world/layout';
import { City, Passport } from '@world/types';
import { ForexHistory, ForexRate } from '@world/types/currency';
import { GoogleRank } from '@world/types/google';
import { Article } from '@world/types/news';
import { YouTubeVideo } from '@world/types/youtube';
import currencyFormatter from '@world/utils/currency-formatter';
import { GetStaticProps, GetStaticPropsContext, NextPage } from 'next';
import Link from 'next/link';
import { CartesianGrid, Legend, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';

const GoogleRanksSection: React.FC<{ googleRanks: GoogleRank[] }> = ({ googleRanks = [] }) => {
  return (
    <>
      {googleRanks.length > 0 ? (
        <TableContainer className="border rounded shadow">
          <Table>
            <Thead>
              <Tr>
                <Th>Google Trends ({googleRanks.length})</Th>
                <Th isNumeric>Occurrences</Th>
              </Tr>
            </Thead>
            <Tbody>
              {googleRanks.map(({ rank, query, count }) => {
                return (
                  <Tr key={rank}>
                    <Td>
                      <Link href={`https://google.com/search?q=${encodeURIComponent(query)}`} target="_blank">
                        <Badge colorScheme="teal">{query}</Badge>
                      </Link>
                    </Td>
                    <Td isNumeric>
                      <b>{count}</b>
                    </Td>
                  </Tr>
                );
              })}
            </Tbody>
            <TableCaption>
              <Link href="/trends" className="uppercase">
                <Button colorScheme="teal" className="w-full mb-4">
                  View Full Table
                </Button>
              </Link>
            </TableCaption>
          </Table>
        </TableContainer>
      ) : (
        <></>
      )}
    </>
  );
};

const NewsSection: React.FC<{ articles: Article[] }> = ({ articles = [] }) => {
  return (
    <>
      {articles.length > 0 ? (
        <div className="grid grid-cols-1 md:grid-cols-2 gap-4 md:gap-8">
          {articles.map(
            ({
              title = '',
              author = '',
              url = '',
              urlToImage = '',
              source: { name: sourceName = '' } = { name: '' },
            }) => {
              return (
                <div key={title} className="col-span-1">
                  <Link href={url || '#'} target="_blank">
                    <Card className="border border-gray-200 overflow-hidden shadow">
                      <div
                        className="aspect-video bg-cover bg-center bg-teal-500"
                        style={{ backgroundImage: `url(${urlToImage || ''})` }}
                      />
                      <CardBody>
                        <div className="flex flex-col gap-2 md:gap-4">
                          <h2 className="font-bold line-clamp-2" title={title}>
                            {title}
                          </h2>
                          <p>{sourceName || author || ''}</p>
                        </div>
                      </CardBody>
                    </Card>
                  </Link>
                </div>
              );
            }
          )}
        </div>
      ) : (
        <></>
      )}
    </>
  );
};

const ForexRates: React.FC<{ forexRates: ForexRate[] }> = ({ forexRates = [] }) => {
  return (
    <>
      {forexRates.length > 0 ? (
        <TableContainer className="border rounded shadow">
          <Table>
            <Thead>
              <Tr>
                <Th>USD</Th>
                <Th isNumeric>1000</Th>
              </Tr>
            </Thead>
            <Tbody>
              {forexRates.map(({ code, rate }) => {
                return (
                  <Tr key={code}>
                    <Td>
                      <Link href={`/currencies/${code}`}>
                        <Badge colorScheme="teal">{code}</Badge>
                      </Link>
                    </Td>
                    <Td isNumeric>{currencyFormatter(rate, code)}</Td>
                  </Tr>
                );
              })}
            </Tbody>
            <TableCaption>
              <Link href="/currencies" className="uppercase">
                <Button colorScheme="teal" className="w-full mb-4">
                  View Full Table
                </Button>
              </Link>
            </TableCaption>
          </Table>
        </TableContainer>
      ) : (
        <></>
      )}
    </>
  );
};

const ForexChart: React.FC<{ forexHistory: ForexHistory[] }> = ({ forexHistory = [] }) => {
  const tos: number[] = forexHistory.map(({ to }) => to);
  const min: number = Math.min(...tos);
  const max: number = Math.max(...tos);
  console.log(min, max);
  return (
    <>
      {forexHistory.length > 0 ? (
        <Card className="border border-gray-200 shadow">
          <CardHeader>
            <Text className="text-center font-bold">USD to EUR</Text>
          </CardHeader>
          <Divider />
          <CardBody>
            <div className="h-[300px] md:h-[450px]">
              <ResponsiveContainer width="100%" height="100%">
                <LineChart width={1600} height={900} data={forexHistory}>
                  <CartesianGrid strokeDasharray="3 3" />
                  <XAxis dataKey="date" />
                  <YAxis domain={[0.9, 0.95]} />
                  <Tooltip />
                  <Legend />
                  <Line type="monotone" dataKey="to" stroke="#82ca9d" />
                </LineChart>
              </ResponsiveContainer>
            </div>
          </CardBody>
        </Card>
      ) : (
        <></>
      )}
    </>
  );
};

const YouTubeVideos: React.FC<{ videos: YouTubeVideo[] }> = ({ videos = [] }) => {
  return (
    <>
      {videos.map(({ id, title, channelTitle, thumbnails }) => {
        const url: string =
          thumbnails?.maxres?.url ||
          thumbnails?.standard?.url ||
          thumbnails?.high?.url ||
          thumbnails?.medium?.url ||
          thumbnails?.default?.url;
        return (
          <div key={id} className="col-span-1">
            <Link href={`https://youtu.be/${id}`} target="_blank">
              <Card className="border border-gray-200 shadow">
                <div
                  className="aspect-video bg-cover bg-center bg-teal-500"
                  style={{ backgroundImage: `url(${url || ''})` }}
                />
                <CardBody>
                  <h2 className="font-bold line-clamp-2" title={title}>
                    {title}
                  </h2>
                  <p>{channelTitle || ''}</p>
                </CardBody>
              </Card>
            </Link>
          </div>
        );
      })}
    </>
  );
};

const PassportsSection: React.FC<{ passports: Passport[] }> = ({ passports = [] }) => {
  return (
    <>
      {passports.length > 0 ? (
        <TableContainer className="border rounded shadow">
          <Table>
            <Thead>
              <Tr>
                <Th>Rank</Th>
                <Th isNumeric>Country</Th>
              </Tr>
            </Thead>
            <Tbody>
              {passports.map(({ countryCode, individualRank, country: { commonName } }) => {
                return (
                  <Tr key={countryCode}>
                    <Td>{individualRank}</Td>
                    <Td isNumeric>
                      <Link href={`/countries/${countryCode}`}>
                        <Badge colorScheme="teal">{commonName}</Badge>
                      </Link>
                    </Td>
                  </Tr>
                );
              })}
            </Tbody>
            <TableCaption>
              <Link href="/passports" className="uppercase">
                <Button colorScheme="teal" className="w-full mb-4">
                  View Full Table
                </Button>
              </Link>
            </TableCaption>
          </Table>
        </TableContainer>
      ) : (
        <></>
      )}
    </>
  );
};

type CountriesPageProps = {
  cities: City[];
  passports: Passport[];
  googleRanks: GoogleRank[];
  forexRates: ForexRate[];
  forexHistory: ForexHistory[];
  articles: Article[];
  videos: YouTubeVideo[];
};

export const CountriesPage: NextPage<CountriesPageProps> = ({
  cities = [],
  googleRanks = [],
  forexRates = [],
  forexHistory = [],
  passports = [],
  articles = [],
  videos = [],
}) => {
  return (
    <Layout>
      <Container>
        <div className="p-4 md:p-8">
          <div className="flex flex-col gap-4 md:gap-8">
            <section className="flex flex-col gap-2 md:gap-4">
              <h1 className="text-2xl uppercase">Weather</h1>
              <Divider />
              <div className="grid grid-cols-1 md:grid-cols-3 gap-4 md:gap-8">
                {cities.map(({ id, city, latitude, longitude, timezone }: City) => (
                  <div key={id} className="col-span-1">
                    <Weather city={city} latitude={latitude} longitude={longitude} timezone={timezone} />
                  </div>
                ))}
              </div>
            </section>
            <section className="flex flex-col gap-2 md:gap-4">
              <h1 className="text-2xl uppercase">Forex</h1>
              <Divider />
              <div className="grid grid-cols-1 md:grid-cols-3 gap-4 md:gap-8">
                <div className="col-span-1 order-2 md:order-1">
                  <ForexRates forexRates={forexRates} />
                </div>
                <div className="col-span-2 order-1 md:order-2">
                  <ForexChart forexHistory={forexHistory} />
                </div>
              </div>
            </section>
            <section className="flex flex-col gap-2 md:gap-4">
              <h1 className="text-2xl uppercase">News</h1>
              <Divider />
              <div className="grid grid-cols-1 md:grid-cols-3 gap-4 md:gap-8">
                <div className="col-span-1">
                  <GoogleRanksSection googleRanks={googleRanks} />
                </div>
                <div className="col-span-2">
                  <NewsSection articles={articles} />
                </div>
              </div>
            </section>
            <section className="flex flex-col gap-2 md:gap-4">
              <h1 className="text-2xl uppercase">YouTube</h1>
              <Divider />
              <div className="grid grid-cols-1 md:grid-cols-3 gap-4 md:gap-8">
                <YouTubeVideos videos={videos} />
              </div>
            </section>
            <section className="flex flex-col gap-2 md:gap-4">
              <h1 className="text-2xl uppercase">Passports</h1>
              <Divider />
              <div className="grid grid-cols-1 md:grid-cols-3 gap-4 md:gap-8">
                <div className="col-span-1">
                  <PassportsSection passports={passports} />
                </div>
                <div className="col-span-2"></div>
              </div>
            </section>
          </div>
        </div>
      </Container>
    </Layout>
  );
};

export const getStaticProps: GetStaticProps = async (
  _context: GetStaticPropsContext
): Promise<{
  props: {
    cities: City[];
    googleRanks: GoogleRank[];
    passports: Passport[];
    articles: Article[];
    forexRates: ForexRate[];
    forexHistory: ForexHistory[];
    videos: YouTubeVideo[];
  };
}> => {
  try {
    const data = await apolloClient.query<{
      cities: City[];
      rates: ForexRate[];
      passports: Passport[];
      history: ForexHistory[];
      google: { ranks: GoogleRank[] };
      news: { headlines: Article[] };
      youtube: { videos: YouTubeVideo[] };
    }>({
      query: HOME_QUERY,
      variables: {
        days: 30,
        from: 'USD',
        to: 'EUR',
        amount: 1000,
        base: 'USD',
        limit: 10,
        pageSize: 4,
        country: 'us',
        maxResults: 6,
      },
    });
    const cities = [...data.data.cities]
      .filter(({ city }) => ['Hà Nội', 'Melbourne', 'Dallas'].includes(city))
      .sort((a, b) => (a.timezone > b.timezone ? 1 : -1));
    const forexRates = [...data.data.rates].splice(0, 10);
    const forexHistory = [...data.data.history];
    const googleRanks = [...data.data.google.ranks];
    const passports = [...data.data.passports];
    const articles = [...data.data.news.headlines];
    const videos = [...data.data.youtube.videos];
    const props = { cities, googleRanks, forexRates, passports, articles, forexHistory, videos };
    console.log('props', props);
    return { props };
  } catch (error) {
    console.error(error);
    return {
      props: { cities: [], googleRanks: [], forexRates: [], passports: [], articles: [], forexHistory: [], videos: [] },
    };
  }
};

export default CountriesPage;
