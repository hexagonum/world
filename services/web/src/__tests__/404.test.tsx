import { render } from '@testing-library/react';
import { NotFoundPage } from '@world/pages/404';

describe('NotFoundPage', () => {
  it('to match snapshot', () => {
    const { container } = render(<NotFoundPage />);
    expect(container).toMatchSnapshot();
  });
});
