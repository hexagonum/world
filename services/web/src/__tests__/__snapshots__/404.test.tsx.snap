// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFoundPage to match snapshot 1`] = `
<div>
  <div
    class="flex flex-col h-screen"
  >
    <nav
      class="border-b shadow"
    >
      <div
        class="container mx-auto"
      >
        <div
          class="px-8 py-4"
        >
          <div
            class="flex items-center justify-between gap-4 lg:gap-8"
          >
            <div
              class="flex items-center gap-4 lg:gap-8"
            >
              <a
                class="text-xl uppercase"
                href="/"
              >
                World
              </a>
              <div
                class="hidden lg:block"
              >
                <div
                  class="flex items-center gap-2 lg:gap-4"
                >
                  <a
                    href="/countries"
                  >
                    Countries
                  </a>
                  <a
                    href="/currencies"
                  >
                    Currencies
                  </a>
                  <a
                    href="/languages"
                  >
                    Languages
                  </a>
                  <a
                    href="/organizations"
                  >
                    Organizations
                  </a>
                  <a
                    href="/passports"
                  >
                    Passports
                  </a>
                  <a
                    href="/timezones"
                  >
                    Timezones
                  </a>
                  <a
                    href="/trends"
                  >
                    Trends
                  </a>
                  <a
                    href="/weather"
                  >
                    Weather
                  </a>
                </div>
              </div>
            </div>
            <div
              class="block lg:hidden"
            >
              <button
                class="chakra-button css-4xx2wk"
                type="button"
              >
                <svg
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </nav>
    <main
      class="grow overflow-hidden"
    >
      <div
        class="h-full overflow-auto"
      >
        <div
          class="flex items-center justify-center h-full"
        >
          <div
            class="text-center"
          >
            <div
              class="text-9xl font-bold"
            >
              404
            </div>
            <div
              class="text-3xl uppercase"
            >
              Page Not Found
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="border-t"
    >
      <div
        class="container mx-auto"
      >
        <div
          class="px-8 py-4"
        >
          <p>
            Â© 
            2023
             
            World
          </p>
        </div>
      </div>
    </footer>
  </div>
</div>
`;
